### ISSUE TITLE
android.lifecycle resolution can be trivially broken by transitive dependencies in Android libraries

### ISSUE THREAD

**Issue by mklim** [flutter/flutter#43383]
Our attempts to solve v2 lifecycle transitive dependency issues are incomplete. There can still trivially be crashes from the wrong version. This was discovered recently by @goderbauer when running [module_test](https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/module_test.dart) failed on stable with the latest plugin versions.

### Background

The v2 Android Embedding brings in multiple dependencies on Android Support libraries. At runtime Gradle will fail to build if a project appears to require different versions of the same library, which is trivial to happen in a Flutter context because plugins can and do include their own versions. In addition to that, plugins that use the v2 embedding _need_ to include their own versions in order to compile against the engine at all.

We are generally trying to work around this by asking Flutter plugins developers to point to a pub package that includes Lifecycle so that they're forced to resolve to the same version. https://pub.dev/packages/flutter_plugin_android_lifecycle

### Immediate problem

Right now module templates created and built with stable will fail to compile if including a v2 embedding plugin because of this issue.

1. `flutter channel stable && flutter create -t module foo`
2. Include any of the migrated flutter/plugins as a dependency in `pubspec.yaml`. `device_info`, `shared_preferences`, etc.
3. `flutter build apk`

The build will fail with `"Android dependency 'android.arch.lifecycle:runtime' has different version for the compile (1.1.0) and runtime (1.1.1) classpath."`.

### Root cause and larger issue

The `module` template is failing on master because its Java code transitively includes a conflicting lifecycle version through `com.android.support:appcompat`.

```
+--- com.android.support:appcompat-v7:27.1.1
|    +--- com.android.support:support-annotations:27.1.1
|    +--- com.android.support:support-core-utils:27.1.1
|    |    +--- com.android.support:support-annotations:27.1.1
|    |    \--- com.android.support:support-compat:27.1.1
|    |         +--- com.android.support:support-annotations:27.1.1
|    |         \--- android.arch.lifecycle:runtime:1.1.0
[...]
\--- project :shared_preferences
|         +--- android.arch.lifecycle:runtime:1.1.1 (*)
```

This is a larger existential issue than just the module template having conflicts and needing to be updated.

- If Java dependencies can cause conflicts, that means that our approach of depending on a pub `flutter_plugin_android_lifecycle` package will not work in all cases. We don't have a current approach to try and prevent this.
- It will be trivially easy and likely for add to app developers to hit conflicts in this way. `com.android.support:appcompat` is a very common dependency for typical Android app development.

### Immediate mitigation

This means that strictly speaking none of the migrated plugins are compatible with stable and should probably be rolled back. /cc @matthew-carroll @amirh If we had a fix we could forward fix them instead, but I don't know if there is an immediate short term one.

### Potential long term fixes

I don't have any solid ideas here yet.

There is a Gradle script to force apps to use a given version for a package. We could add that in, but I don't like the idea of deciding dependency resolution in all cases like this.

/cc @blasten

**Comment by mklim** [comment]
Also /cc @tvolkert @csells

**Comment by matthew-carroll** [comment]
Isn't this a standard compatibility issue? If any other app were using this version of appcompat, and a library that uses Lifecycle 1.1.1, wouldn't this same failure occur?

**Comment by mklim** [comment]
Yeah, the technical problem is the same. I think our overall ecosystem makes it worse than it generally is with typical Android development though.

- Android developers write their own Gradle files to start with. Flutter developers can also edit their Gradle scripts to force version resolution if needed, but it's generally much worse for them to have to manage these files manually, and figuring out how to solve this problem initially may require them to ramp up on Gradle for the first time.
- I don't think it's as common for Android app developers to depend on libraries that then use android.support/androidx libraries as transitive dependencies, outside of the android.support/androidx libraries themselves. I could be wrong on this though, it's just my own rough personal impression. 
- Presumably in the Android ecosystem developers can address this by changing the support version that their app uses, or going to an earlier version of the offending library that matches their dependency. In the Flutter ecosystem we are going to be switching to just one version and requiring users to change their app to match that, which may not be feasible in an a2a scenario. This could happen in the general Android ecosystem for a given library too, but here we would be requiring it for all packages with the v2 embedding.

**Comment by matthew-carroll** [comment]
Update after some discussion:

- The Fragment dependency is a possible issue, too. We think that the Lifecycle problem is broad enough, that solving just for Lifecycle is worth the effort. Fragment conflicts will still need to be resolved by changing dependency versions in the final app. We still need to validate this assumption.
- We think that hiding Lifecycle specifically from the plugin APIs will solve the problem for the ecosystem. We still need to validate this assumption.
- If both above assumptions are correct, then we can produce a Flutter version of Lifecycle to break the deps
- If all above are done, do we want to extend the new plugin to include an adapter that converts Flutter Lifecycle to real Lifecycle?

**Comment by tvolkert** [comment]
It seems like there's several different points we've run into (this being the most recent) that argue for a level of indirection between android support (AndroidX) classes and Flutter equivalents, where we provide adapters to translate between the two.

**Comment by mklim** [comment]
From some quick testing offline, Including `lifecycle` in the plugins via `compileOnly` avoids this conflict. Right now we're including them via `api`. `implementation` still hits this issue. `compileOnly` does push the burden onto the user for including the library if the plugins actually needed to use it at runtime, but so far I don't think any of them do. This will probably be an issue for Fragment plugins though.

We should forward fix the plugins to include these deps via `compileOnly` for now.

**Comment by xster** [comment]
I tried playing around with this a little bit and the more I look at this, the more I'm thinking "I don't think there's really an actual bug here".

Let's split this into 4 parts 

- full-flutter on stable
- add-to-app on stable
- full-flutter on master
- add-to-app on master

### Add-to-app on stable

I think this is the main concrete problem this issue is pointing to (since the repro is with a module template). 

This isn't a "real" issue on a few levels. 

1.  add-to-app isn't "shipped" on stable. We're not expecting anyone to "use" add-to-app with transitive dependencies unless they're recent enough on the tree to get the engine POM (which brings in the Lifecycle and doesn't need the plugins to bring the Lifecycles in as a temporary stopgap ) or they're already manually constructing their .gradle and fulfilling Lifecycle for us manually (in which case they're already taking everything into their own hands.

2. it's a bit confusing currently until we solve https://github.com/flutter/flutter/issues/37910 but `flutter build apk` inside a module isn't particularly meaningful. A module has both a module part (very important) and a lightweight wrapper app part (not as important). The problem arises from us not having super cleanly maintaining the wrapper app part's template. It's even a bit internally self-contradictory (using [compile SDK 28](https://github.com/flutter/flutter/blob/68587a0916366e9512a78df22c44163d041dd5f3/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L6) and [appcompat 27](https://github.com/flutter/flutter/blob/68587a0916366e9512a78df22c44163d041dd5f3/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L50)). 

3. it's not really an ecosystem problem but an entirely self-inflicted problem since in our own mono-framework-tools-engine repo, we control the version of this appcompat ([in the ephemeral auto-generated app gradle](https://github.com/flutter/flutter/blob/68587a0916366e9512a78df22c44163d041dd5f3/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L50) (which we've also since fixed on master in [#39798]( https://github.com/flutter/flutter/pull/39798/files#diff-ed21264ed964a2dc344e9ae9af805356R50)) which we set to v27 as well as the version of the lifecycle ([in the json file which populates the engine POM](https://github.com/flutter/engine/blob/master/tools/android_support/files.json) and the temporary plugin repo's afterEvaluate workaround). 

### Full-flutter on stable

1. this seems like a self-created problem due to using "api" on the plugins as the [temporary stopgap](https://github.com/flutter/flutter/issues/42348) and shouldn't affect our final POM-based solution. 

2. normal full-flutter apps shouldn't have conflicting versions of Lifecycle since they have barebones Android code. That is unless they "add-to-Flutter" and augment the Android template or bring in other conflicting plugins manually. At which point we're at the present state of how existing plugins are with conflicting transitive native dependencies (until it's fixed by Collin's firebase PR https://github.com/FirebaseExtended/flutterfire/pull/39). 

3. we can fix them the same way with Michael's PR https://github.com/flutter/plugins/pull/2232. It seems like the correct fix since there's no reason to re-export the Lifecycle from the plugins. 

### Add-to-app / full-flutter on master

This was my primary concern but as Emmanuel mentioned in email, it seems to be not as fragile as we thought. 

https://github.com/xster/flutter-test/tree/master/t136_transitive_dependency_conflict 

and 

https://github.com/xster/flutter-test/tree/master/t136_host_sub_module

are a typical pairing of a Flutter module with an engine POM and an Android app. 

Even if the Android app produces a conflicting native dependency to Flutter engine POM's transitive native dependencies (such as referencing appcompat v27 in this test https://github.com/xster/flutter-test/blob/f030a6cf5c7c7e3b6a2dfd06bcdf550166bde104/t136_host_sub_module/app/build.gradle#L33), gradle can still self-resolve the versioning conflict correctly. 

See https://scans.gradle.com/s/djd6spd34tqtc/dependencies?expandAll&focusedDependency=WzAsMTQsMTUsWzAsMTQsWzEzLDE1XV1d&focusedDependencyView=details&toggled=W1swLDNdLFswLDRdLFswLDVdLFswLDZdLFswLDddLFswLDE0LFsxM11dLFswLDE0LFsxMywxNV1dLFswLDE0LFsxMywxNSwxN11dLFswLDE0LFs5OF1dLFswLDE0LFs5OCwxODVdXV0

<img width="867" alt="Screen Shot 2019-10-25 at 5 17 34 AM" src="https://user-images.githubusercontent.com/156888/67570712-ec5dec00-f6e6-11e9-8462-c652da782606.png">

The engine POM's android.arch.lifecycle:runtime:1.1.1 (in red) is theoretically in conflict with the app's self-declared com.android.support:appcompat-v7:27.1.1 (in blue), but gradle self adjusted appcompat's upstream dependency support-core-utils (in green) to v28 instead of v27. 

### TL;DR

The repro here specifically points to issues with the Lifecycle work-around. I think the existing general long-term POM-based Lifecycle solution might still be ok. 

If future users manages to combine transitive dependencies in ways that are so incompatible that it can't build, we probably can't do much better than pointing users to https://docs.gradle.org/current/userguide/managing_transitive_dependencies.html for manual version resolution which is more or less the present state of the ecosystem (https://github.com/flutter/flutter/issues/27254).

**Comment by mklim** [comment]
> I'm thinking "I don't think there's really an actual bug here".

Thank you for the detailed investigation and analysis @xster, I think your conclusion makes sense. I do have one last breakage that I'm not totally making sense of.

Adding an incompatible version of lifecycle to an AndroidX app on master will still cause build failures today, mklim/test_app@69fd6875eae796898d05f273b26b76b7b91dbf50. That shouldn't be happening if Gradle is version solving, right?

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:compileReleaseJavaWithJavac'.
> Could not resolve all task dependencies for configuration ':app:releaseCompileClasspath'.
   > Could not find androidx.lifecycle:lifecycle-runtime:2.2.0.
     Required by:
         project :app
   > Could not find androidx.lifecycle:lifecycle-runtime:2.2.0.
     Required by:
         project :app > io.flutter:flutter_embedding_release:1.0.0-2abf1101b482932f68533061b07f492ef203052c
         project :app > androidx.core:core:1.0.0
```

Changing that patch to use `2.1.0` "fixes" it. `2.1.0` is the exact version that Jetifier will convert the engine's original `android.arch` dependency to. Using either `android.arch` dependency on a non-AndroidX app appears to work correctly like you've found. The plugins are AndroidX in general so I think this was confounding our test cases so far.

Is it maybe that Gradle isn't solving for y increments or some other scheme where `2.2.0` happens to be considered breaking? That or maybe Jetifier is interfering. I'm looking into how Gradle considers versions to be "too high" to try and get to the bottom of it.

**Comment by amirh** [comment]
>  The engine POM's android.arch.lifecycle:runtime:1.1.1 (in red) is theoretically in conflict with the app's self-declared com.android.support:appcompat-v7:27.1.1 (in blue), but gradle self adjusted appcompat's upstream dependency support-core-utils (in green) to v28 instead of v27. 

It seems like appcompat 28 and 27 both depend on lifecycle 1.1.x which is compatible with the lifecycle version that the engine wants, how gradle handles this if these were incompatible versions of lifecycle? (e.g if one of the engine or app transitive dependencies was lifecycle 2.x.y)

**Comment by blasten** [comment]
> It seems like appcompat 28 and 27 both depend on lifecycle 1.1.x which is compatible with the lifecycle version that the engine wants, how gradle handles this if these were incompatible versions of lifecycle? (e.g if one of the engine or app transitive dependencies was lifecycle 2.x.y)

Gradle will pick the largest version. 

I tested this by bringing these two dependencies (one if the app, the other one in a plugin):
```gradle
  implementation 'org.webjars.npm:dompurify:1.0.4'
```

and
```gradle
  implementation 'org.webjars.npm:dompurify:2.0.7'
```

Gradle resolved `dompurify` to `2.0.7`. 

This is surprising to me. I will ask on the Gradle repo.

**Comment by xster** [comment]
> It seems like appcompat 28 and 27 both depend on lifecycle 1.1.x which is compatible with the lifecycle version that the engine wants

Can you elaborate on how you determined that?

It seems like looking at com.android.support:appcompat-v7:27.1.1 in https://mvnrepository.com/artifact/com.android.support/appcompat-v7/27.1.1 and its POM file, it specifically has 

```
<dependency>
  <groupId>com.android.support</groupId>
  <artifactId>support-core-utils</artifactId>
  <version>27.1.1</version>
  <type>aar</type>
  <scope>compile</scope>
</dependency>
```

which references https://mvnrepository.com/artifact/com.android.support/support-core-utils/27.1.1 whose POM has

```
<dependency>
  <groupId>com.android.support</groupId>
  <artifactId>support-compat</artifactId>
  <version>27.1.1</version>
  <type>aar</type>
  <scope>compile</scope>
</dependency>
```

which references https://mvnrepository.com/artifact/com.android.support/support-compat/27.1.1 that specifically points to 

```
<dependency>
  <groupId>android.arch.lifecycle</groupId>
  <artifactId>runtime</artifactId>
  <version>1.1.0</version>
  <type>aar</type>
  <scope>compile</scope>
  <exclusions>
    <exclusion>
      <artifactId>support-annotations</artifactId>
      <groupId>*</groupId>
    </exclusion>
  </exclusions>
</dependency>
```

which is lifecycle 1.1.0.

**Comment by amirh** [comment]
> Gradle will pick the largest version.

Can this lead to compile/runtime errors?

**Comment by amirh** [comment]
> > It seems like appcompat 28 and 27 both depend on lifecycle 1.1.x which is compatible with the lifecycle version that the engine wants
> 
> Can you elaborate on how you determined that?
> ...
> which is lifecycle 1.1.0.

Isn't that the same determination?

**Comment by blasten** [comment]
I filed this issue: https://github.com/gradle/gradle/issues/11144. Let me also send it over to the internal Gradle chat.

**Comment by blasten** [comment]
> Can this lead to compile/runtime errors?

Yes. Unless version constraints are manually specified.  In Gradle 6.0 (rc), there will be a [new artifact](https://docs.gradle.org/6.0-rc-1/userguide/publishing_gradle_module_metadata.html) that can be published along with the pom file to specify fined grained version constraints.

**Comment by xster** [comment]
Ah ok. I think you're asking the reverse question of what this issue is about. 

This issue was bringing up a potentially new issue about how everything could be potentially broken all the time because it's super easy to transitively depend on lifecycle multiple times and the versions will be super fragile and the build will break too frequently. I was saying that this new issue isn't an issue. 

Since the conflict resolution isn't as fragile as we thought yesterday, your current question is will the build fail too infrequently now. e.g. 

> how gradle handles this if these were incompatible versions of lifecycle? (e.g if one of the engine or app transitive dependencies was lifecycle 2.x.y)

I think that's the question we always knew and talked about it not being possibly magically solvable. It sounds like POM will never crash unless we use <dependencyManagement> in POM or dependencies.constraints in gradle (none of which we do). Tested this by depending on com.android.support:appcompat-v7:27.0.0 which brings in lifecycle 1.0.0 and directly pointing to lifecycle 1.0.0 directly in gradle. 

In other words, this will crash at compile time if there was a backward-non-compatible API change rather than at dependency resolving phase. 

I think that's what you were pushing for yesterday since this would be the same behavior if we can hypothetically hide lifecycle behind our own wrapper and make versions implicit.

**Comment by xster** [comment]
Can confirm that with @mklim's fix, device_info 0.4.1 crashes with the originally documented error and 0.4.1+1 will build fine.

**Comment by mklim** [comment]
@matthew-carroll asked if we could try summarizing the known and previously discussed issues in a doc. I took an attempt at it [in this doc](https://docs.google.com/document/d/1s_Jol7kEWel8driKZObjcf5_CVFk6VH5jJJ4Mw6_Ceo/edit). We cycled on what the issues really were a few times so it could be missing things, feel free to comment or edit it out more.

**Comment by mklim** [comment]
I think we've come to a consensus and resolved this issue as much as possible, so I'm going to close it.

1. The error noted in the original comment can be trivially resolved on the plugin side and doesn't have wider implications.
2. We're moving forward with the pub package as a sign of the underlying Java dependency. This doesn't enforce a true dependency as uncovered in this bug but it's our best option for now.
3. Once Gradle supports it we'll start enforcing a range of dependencies in the Flutter engine and link that to the pub package.

Feel free to reopen it if I'm missing something and there's more work to do here.
